src/resmom/renew.c

```diff
@@ -0,0 +1,1219 @@
+ *	via IM protocol. This function is meant to by run by send_sisters and
+ *	it shouldn't be sent via mcast because mcast can't be wrapped by GSS
+ *
+ * @param[in] jobid - Job ID
+ *
+ * @return 	int
+ * @retval	DIS_SUCCESS on success
+ * @retval	!= DIS_SUCCESS otherwise
+ */
+int
+im_cred_send(job *pjob, hnodent *xp, int stream)
+{
+	int		ret;
+	svrcred_data	*cred_data;
+	char		*data_base64;
+
+	cred_data = find_cred_data_by_jobid(pjob->ji_qs.ji_jobid);
+
+	if (cred_data == NULL || (data_base64 = cred_data->cr_data_base64) == NULL) {
+		ret = DIS_PROTO;
+		goto done;
+	}
+
+	ret = diswui(stream, cred_data->cr_type);
+	if (ret != DIS_SUCCESS)
+		goto done;
+
+	ret = diswst(stream, data_base64);
+	if (ret != DIS_SUCCESS)
+		goto done;
+
+	ret = diswul(stream, cred_data->cr_validity);
+	if (ret != DIS_SUCCESS)
+		goto done;
+
+	return DIS_SUCCESS;
+
+done:
+	sprintf(log_buffer, "dis err %d (%s)", ret, dis_emsg[ret]);
+	DBPRT(("%s: %s\n", __func__, log_buffer))
+	log_joberr(-1, __func__, log_buffer, pjob->ji_qs.ji_jobid);
+	return ret;
+}
+
+/**
+ * @brief
+ * 	im_cred_read - read received (via IM) credentials in base64 on sister
+ *	mom, store the credentials in mom's memory and renew the credentials for
+ *	associated job.
+ *
+ * @param[in] pjob - job structure
+ * @param[in] np - node structure
+ * @param[in] stream - tpp channel
+ *
+ * @return 	int
+ * @retval	DIS_SUCCESS on success
+ * @retval	!= DIS_SUCCESS otherwise
+ */
+int
+im_cred_read(job *pjob, hnodent *np, int stream)
+{
+	int		ret;
+	char		*data_base64;
+	unsigned char	out_data[CRED_DATA_SIZE];
+	ssize_t		out_len = 0;
+	char		buf[LOG_BUF_SIZE];
+	krb5_data	*data;
+	int		cred_type;
+	long		validity;
+
+	DBPRT(("%s: entry\n", __func__))
+
+	cred_type = disrui(stream, &ret);
+	if (ret != DIS_SUCCESS)
+		goto err;
+
+	data_base64 = disrst(stream, &ret);
+	if (ret != DIS_SUCCESS)
+		goto err;
+
+	validity = disrul(stream, &ret);
+	if (ret != DIS_SUCCESS)
+		goto err;
+
+	if (decode_block_base64((unsigned char *)data_base64, strlen(data_base64), out_data, &out_len, buf, LOG_BUF_SIZE) != 0) {
+		log_err(errno, __func__, buf);
+		ret = DIS_PROTO;
+		goto err;
+	}
+
+	if ((data = (krb5_data *)malloc(sizeof(krb5_data))) == NULL) {
+		log_err(errno, __func__, "Unable to allocate Memory!\n");
+		ret = DIS_NOMALLOC;
+		goto err;
+	}
+
+	if ((data->data = (char *)malloc(sizeof(unsigned char)*out_len)) == NULL) {
+		log_err(errno, __func__, "Unable to allocate Memory!\n");
+		ret = DIS_NOMALLOC;
+		goto err;
+	}
+
+	data->length = out_len;
+	memcpy(data->data, out_data, out_len);
+
+	log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_JOB,
+		LOG_INFO,
+		pjob->ji_qs.ji_jobid,
+		"credentials from superior mom received");
+
+	store_or_update_cred(pjob->ji_qs.ji_jobid,
+		pjob->ji_wattr[(int)JOB_ATR_cred_id].at_val.at_str,
+		cred_type,
+		data,
+		NULL,
+		validity);
+
+	/* I am the sister and new cred has been received - lets renew creds */
+	renew_job_cred(pjob);
+
+	return DIS_SUCCESS;
```

Comment:
dont we need to free() data and data->data here?

Reply:
No, store_or_update_cred() stores both in the memory and delete_cred() will free both later.